File: apps/desktop/src-tauri/src/cli_install.rs
//! CLI 佈署的殼層（desktop-app spec「安裝 CLI 指令到 PATH」，design D5）：
//! 收集環境事實（平台、PATH、sidecar 路徑、已佈署版本輸出）與執行佈署計畫
//! （symlink／copy）。狀態判定與平台分流的決策歸前端 core（vitest 覆蓋），
//! 此處不做任何判斷。

use std::path::PathBuf;
use std::process::Command;

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CliInstallProbe {
    pub platform: String,
    pub home: Option<String>,
    pub path_env: String,
    pub path_delimiter: String,
    pub bundled_cli_path: Option<String>,
    pub app_version: String,
    pub deployed_version_output: Option<String>,
}

fn platform_key() -> String {
    if cfg!(target_os = "macos") {
        "macos".into()
    } else if cfg!(target_os = "windows") {
        "windows".into()
    } else if std::env::var_os("APPIMAGE").is_some() {
        "linux-appimage".into()
    } else {
        "linux-deb".into()
    }
}

/// 隨附 CLI 的位置：Tauri externalBin 佈於主執行檔同目錄。
fn bundled_cli_path() -> Option<PathBuf> {
    let exe = std::env::current_exe().ok()?;
    let name = if cfg!(target_os = "windows") { "speclink.exe" } else { "speclink" };
    let path = exe.parent()?.join(name);
    path.exists().then_some(path)
}

/// 已佈署 CLI 的 `--version` 原始輸出（解析歸前端 core）；不存在／不可執行＝None。
fn deployed_version_output(platform: &str, home: Option<&str>) -> Option<String> {
    let program: PathBuf = match platform {
        "macos" | "linux-appimage" => PathBuf::from(home?).join(".local/bin/speclink"),
        "linux-deb" => PathBuf::from("/usr/bin/speclink"),
        // Windows 由安裝器把 CLI 佈於 app 同目錄並寫 PATH；剛裝完時 PATH
        // broadcast 未及已執行程序，走程序 PATH 會誤報未安裝——直接以
        // sidecar 同目錄（即 $INSTDIR）執行。
        _ => bundled_cli_path()?,
    };
    let output = Command::new(program).arg("--version").output().ok()?;
    output.status.success().then(|| String::from_utf8_lossy(&output.stdout).into_owned())
}

/// PATH 提示的判定基準是「使用者終端」的 PATH——macOS GUI app 從 Dock 啟動
/// 不繼承 shell 設定檔，程序 PATH 只有系統預設，直接用會誤報「不在 PATH」。
/// 以 login shell 解析真實 PATH；失敗（非 unix、shell 異常）退回程序 PATH。
fn user_shell_path() -> String {
    #[cfg(unix)]
    {
        let shell = std::env::var("SHELL").unwrap_or_else(|_| "/bin/sh".into());
        if let Ok(output) = Command::new(&shell).args(["-lc", r#"printf %s "$PATH""#]).output() {
            if output.status.success() {
                let path = String::from_utf8_lossy(&output.stdout).into_owned();
                if !path.trim().is_empty() {
                    return path;
                }
            }
        }
    }
    std::env::var("PATH").unwrap_or_default()
}

pub fn probe(app_version: String) -> CliInstallProbe {
    let platform = platform_key();
    let home = std::env::var("HOME")
        .ok()
        .or_else(|| std::env::var("USERPROFILE").ok());
    let deployed_version_output = deployed_version_output(&platform, home.as_deref());
    CliInstallProbe {
        platform,
        path_env: user_shell_path(),
        path_delimiter: if cfg!(target_os = "windows") { ";".into() } else { ":".into() },
        bundled_cli_path: bundled_cli_path().map(|p| p.to_string_lossy().into_owned()),
        app_version,
        deployed_version_output,
        home,
    }
}

/// 前端 core 產出的佈署計畫（「none」不會被送來——前端已擋）。
#[derive(serde::Deserialize)]
#[serde(tag = "action", rename_all = "camelCase")]
pub enum CliDeployPlan {
    #[serde(rename = "symlink", rename_all = "camelCase")]
    Symlink { link_path: PathBuf, target_path: PathBuf },
    #[serde(rename = "copy", rename_all = "camelCase")]
    Copy { dest_path: PathBuf, source_path: PathBuf },
}

pub fn deploy(plan: CliDeployPlan) -> Result<(), String> {
    match plan {
        CliDeployPlan::Symlink { link_path, target_path } => {
            let parent = link_path.parent().ok_or("symlink 路徑缺少父目錄")?;
            std::fs::create_dir_all(parent).map_err(|e| e.to_string())?;
            // 既有 symlink／檔案先移除，重複安裝與換目標皆冪等。
            if std::fs::symlink_metadata(&link_path).is_ok() {
                std::fs::remove_file(&link_path).map_err(|e| e.to_string())?;
            }
            #[cfg(unix)]
            return std::os::unix::fs::symlink(&target_path, &link_path).map_err(|e| e.to_string());
            #[cfg(not(unix))]
            {
                let _ = target_path;
                Err("此平台不支援 symlink 佈署".into())
            }
        }
        CliDeployPlan::Copy { dest_path, source_path } => {
            let parent = dest_path.parent().ok_or("複製目的地缺少父目錄")?;
            std::fs::create_dir_all(parent).map_err(|e| e.to_string())?;
            std::fs::copy(&source_path, &dest_path).map_err(|e| e.to_string())?;
            #[cfg(unix)]
            {
                use std::os::unix::fs::PermissionsExt;
                std::fs::set_permissions(&dest_path, std::fs::Permissions::from_mode(0o755))
                    .map_err(|e| e.to_string())?;
            }
            Ok(())
        }
    }
}

File: apps/desktop/src/core/cliInstall.ts
// CLI 佈署純邏輯（desktop-app spec「安裝 CLI 指令到 PATH」，design D5）：狀態
// 判定與平台分流的佈署計畫。檔案系統與子程序歸 Tauri 殼（收集事實、執行計畫），
// 這裡只做決策——vitest 可測、不依賴 Tauri。
export type CliPlatform = "macos" | "windows" | "linux-appimage" | "linux-deb";

export type CliInstallStatus =
  | { kind: "not-installed" }
  | { kind: "installed"; version: string }
  | { kind: "version-mismatch"; version: string };

export type CliDeployPlan =
  | { action: "symlink"; linkPath: string; targetPath: string }
  | { action: "copy"; destPath: string; sourcePath: string }
  | { action: "none"; reason: "installer-managed" | "package-managed" };

/** 以偵測到的已佈署版本（null＝未偵測到）對 app 版本判定三態。 */
export function cliInstallStatus(
  deployedVersion: string | null,
  appVersion: string,
): CliInstallStatus {
  if (deployedVersion === null) return { kind: "not-installed" };
  if (deployedVersion === appVersion) return { kind: "installed", version: deployedVersion };
  return { kind: "version-mismatch", version: deployedVersion };
}

/** 解析 `speclink --version` 輸出的套件版號（如「speclink 0.1.0 (arm64, engine
 * v1.14.0)」取 0.1.0；括號內的引擎版號不參與佈署判定）；不合形回 null。 */
export function parseCliVersion(output: string): string | null {
  return output.match(/^speclink (\S+)/)?.[1] ?? null;
}

/** 平台分流的佈署計畫（design D5）：macOS symlink（app 路徑固定、更新後自動指向
 * 新版）；AppImage 複製（掛載點隨執行變動、symlink 不可行）；Windows 由 NSIS
 * 安裝器寫 PATH、deb 由包管理器佈署 /usr/bin——後兩者 app 內僅回報狀態。 */
export function cliDeployPlan(
  platform: CliPlatform,
  ctx: { home: string; bundledCliPath: string },
): CliDeployPlan {
  switch (platform) {
    case "macos":
      return {
        action: "symlink",
        linkPath: `${ctx.home}/.local/bin/speclink`,
        targetPath: ctx.bundledCliPath,
      };
    case "linux-appimage":
      return {
        action: "copy",
        destPath: `${ctx.home}/.local/bin/speclink`,
        sourcePath: ctx.bundledCliPath,
      };
    case "windows":
      return { action: "none", reason: "installer-managed" };
    case "linux-deb":
      return { action: "none", reason: "package-managed" };
  }
}

/** 啟動自動佈署判定（spec 修訂「macOS 啟動自動佈署與自我修復」＋「AppImage
 * 版本不符自我修復」，design D12）：macOS 未安裝或版本不符皆自動佈署（symlink
 * 冪等無副作用）；AppImage 僅版本不符時修復——未安裝仍是使用者的顯式動作
 * （複製佈署較重）；Windows／deb 歸安裝器與包管理器。 */
export function needsRedeploy(platform: CliPlatform, status: CliInstallStatus): boolean {
  if (platform === "macos") return status.kind !== "installed";
  return platform === "linux-appimage" && status.kind === "version-mismatch";
}

/** ~/.zprofile 的識別註解：追加行帶著它，冪等判定靠它。 */
export const ZPROFILE_MARKER = "# speclink cli";

export type ZprofilePlan = { action: "append"; line: string } | { action: "none" };

/** macOS PATH 追加決策（design D12）：已在 PATH 或 zprofile 已有識別行→不寫；
 * 否則輸出帶識別註解的一行（zprofile 讀不到視為空檔）。寫檔歸 adapter。 */
export function zprofilePlan(onPath: boolean, zprofile: string | null): ZprofilePlan {
  if (onPath) return { action: "none" };
  if ((zprofile ?? "").includes(ZPROFILE_MARKER)) return { action: "none" };
  return { action: "append", line: `export PATH="$HOME/.local/bin:$PATH" ${ZPROFILE_MARKER}` };
}

/** 佈署目錄是否已在 PATH 上；否則介面提示加入方式。 */
export function isDirOnPath(dir: string, pathEnv: string, delimiter: ":" | ";"): boolean {
  return pathEnv.split(delimiter).includes(dir);
}
